Add support for TPM- and Tang-based disk encryption

simoncampion opened the pull request:
# Add support for TPM- and Tang-based disk encryption

This PR adds support for TPM- and Tang-based disk encryption, by adding an `app-crypt/clevis` package and dependencies and an updated version of `sys-kernel/bootengine`. It is accompanied by PRs in the [bootengine](https://github.com/flatcar/bootengine/pull/83) and [mantle](https://github.com/flatcar/mantle/pull/493) repositories. It includes work by @krishjainx  towards supporting these features, plus a few further changes that I needed to make for the features to fully work. I left changes by @krishjainx largely untouched where no modifications were needed to make it work, but I'm happy to address comments about any of the changes in this PR.

## How to use

With these changes the `luks.clevis.tpm2` and `luks.clevis.tang` Ignition options should be fully supported. See the automated tests in the [mantle pull request](https://github.com/flatcar/mantle/pull/493) for examples.

I have no prior experience making changes to Flatcar and look forward to hearing how to do things better. One specific aspect that I'd like feedback on is the following issue: When encrypting the ROOT partition, it must be decrypted in the initramfs before the root pivot. However, Ignition will add an `/etc/crypttab` entry for the ROOT partition, leading to the generation of a `systemd-cryptsetup` service for the encrypted root. Unfortunately, due to how this service is configured, it will fail if the root disk is already decrypted. This is a [well-known issue](https://unix.stackexchange.com/questions/275734/why-does-systemd-cryptsetup-try-to-remount-the-root-partition-already-mounted). This does not impact the correct functioning of the system, but is annoying and also leads to the kola tests for root disk encryption to fail since the test system always checks for failed services. I'm unsure how to best fix this issue and would be thankful for suggestions.

## Testing done

I tested the `luks.clevis.tpm2` and `luks.clevis.tang` config options for the ROOT partition and non-root disks using the new tests in kola.

- [ ] Changelog entries added in the respective `changelog/` directory (user-facing change, bug fix, security fix, update)
- [ ] Inspected CI output for image differences: `/boot` and `/usr` size, packages, list files for any missing binaries, kernel modules, config files, kernel modules, etc.

<!-- For coreos-overlay ebuild modifications that include a CROS_WORKON_COMMIT bump, did you bump too the ebuild revision ? -->

Once this is merged, I'd also be happy to update the documentation and create a new Butane spec for the newly supported config options.

pothos commented:
Thanks!
Let's keep the rootfs out of scope for now because it will also be tricky to get the dependencies right.

When a package from Gentoo needs no changes, it should be in portage-stable. The profile file can be used for simple tweaks. If the package needs changes, we have it in coreos-overlay but in two commits: the first "syncs from Gentoo" which is deleting any files in the package folder and brings in new upstream files, and the second commit "applies Flatcar changes" and this is where commented changes are introduced. Only then we can correctly update the package.

simoncampion commented:
Thanks for the speedy comment, @pothos !

> When a package from Gentoo needs no changes, it should be in portage-stable. The profile file can be used for simple tweaks. If the package needs changes, we have it in coreos-overlay but in two commits: the first "syncs from Gentoo" which is deleting any files in the package folder and brings in new upstream files, and the second commit "applies Flatcar changes" and this is where commented changes are introduced. Only then we can correctly update the package.

That makes sense :+1: I reworked the Git history accordingly and cleaned things up a bit. Let me know if there's more to change; I did my best to infer how things are done from existing commits, but might have missed some bits.

If I see correctly, Krish found some of the packages we need (`app-crypt/clevis`, `dev-libs/jose`, and `dev-libs/luksmeta`) in GURU rather than the official Gentoo repo. I used the `from Gentoo` commits for these packages too. If you prefer to treat packages from GURU differently in terms of how they are committed, let me know.

simoncampion commented:
Okay, I moved the packages from GURU to `coreos-overlay`.

@pothos If there is anything else I can do for this PR, let me know.

pothos commented on sdk_container/src/third_party/coreos-overlay/app-crypt/clevis/metadata.xml:
```diff
@@ -2,14 +2,15 @@
 	<use>
 		<flag name="luks">Enable LUKS support</flag>
 		<flag name="tpm">Enable TPM support</flag>
+		<flag name="dracut">Enable dracut support</flag>
```
I don't think we need a new use flag here but can directly depend on dracut

simoncampion replied:
Sounds good. I removed the use flag.

pothos commented:
Looks ok but needs some cleanup: All unmodified ebuild folders should go under `portage-stable` not `coreos-overlay`, e.g., jose.

simoncampion commented:
> Hello @simoncampion, can you please rebase on top of flatcar/scripts main branch so that we can trigger a CI run too?
> 
> Thanks.

@ader1990 Will do! I suggest we wait for the associated changes of the [bootengine](https://github.com/flatcar/bootengine/pull/83) to be finalized. I'll then update this PR to point to the correct version of `bootengine`, move unmodified ebuilds from `coreos-overlay` to `portage-stable`, as suggested by @pothos , and rebase on master.

simoncampion commented:
> Looks ok but needs some cleanup: All unmodified ebuild folders should go under `portage-stable` not `coreos-overlay`, e.g., jose.

I moved all unmodified ebuild folders (all but `app-crypt/clevis`) to `portage-stable`.

I also rebased on `main`. I'll build this tomorrow morning locally and check the initramfs size and make sure all disk encryption tests still pass.

pothos commented on sdk_container/src/third_party/coreos-overlay/app-crypt/clevis/files/clevis-dracut.patch:
```diff
@@ -0,0 +1,216 @@
+new file mode 100755
+index 0000000..a3b9d62
+--- /dev/null
++++ b/src/dracut/clevis/clevis-luks-generic-unlocker
```
Do we make us of `clevis-luks-generic-unlocker` or is this not needed anymore?

simoncampion replied:
I think we do not need `clevis-luks-generic-unlocker` anymore.

However, I wonder whether we should remove it here. This `.patch` file, which creates `clevis-luks-generic-unlocker`, is part of the upstream ebuild from GURU, so I thought that it'd be preferable to keep it untouched, to make it easier to update in the future? I guess we could remove `clevis-luks-generic-unlocker` in our `.patch` file which is applied after the upstream `.patch` file, but I wonder why we would do that. `clevis-luks-generic-unlocker` shouldn't be included in our initramfs image anyways because of the logic [here](https://github.com/flatcar/scripts/pull/1560/commits/406ee4c085907fe0fe433c07a0a45c5bb55bf750#diff-b2d6a79ae4b1b7c7a321cf35f7bd69b16a287465501162f1bfb2481604e19ae2R146).

(I did see though that we still create a symlink under `/usr/bin` in the initramfs image, pointing to `clevis-luks-generic-unlocker` on /sysusr/usr. I think that symlink is not needed and I will remove it in the bootengine PR.)

pothos replied:
Ah, ok, sure, it can stay, thanks for the info.

pothos commented:
The CI failed with `emerge: there are no ebuilds to satisfy "sys-devel/autoconf-archive".` - maybe a fallout from the rebase on `main`?

pothos commented on sdk_container/src/third_party/coreos-overlay/sys-kernel/bootengine/bootengine-9999.ebuild:
```diff
@@ -2,15 +2,15 @@
 EAPI=7
-CROS_WORKON_PROJECT="flatcar/bootengine"
+CROS_WORKON_PROJECT="simoncampion/bootengine"
 CROS_WORKON_LOCALNAME="bootengine"
 CROS_WORKON_OUTOFTREE_BUILD=1
 CROS_WORKON_REPO="https://github.com"
 
 if [[ "${PV}" == 9999 ]]; then
 	KEYWORDS="~amd64 ~arm ~arm64 ~x86"
 else
-	CROS_WORKON_COMMIT="0bade95d3b33b75b6c827d2db2f9298aff0ca05f" # flatcar-master
+	CROS_WORKON_COMMIT="32419d727df92c9212b9b7405ccb95922c8fd4d4"
```
The PR is merged, we can use the merge commit now, and then we also need a changelog entry in `changelog/changes/` like `- Added Ignition Clevis support for encrypted root disks unlocked with a TPM2 device or a tang-server ([scripts#1560](https://github.com/flatcar/scripts/pull/1560))`

simoncampion replied:
I updated it to refer to the merge commit, and I also added a changelog entry.

ader1990 commented:
> Before we hit merge, I'd like to see the new tests for disk encryption pass on the most recent version of this PR. The QEMU image I built locally hangs in boot, failing to modprobe btrfs ("Too many levels of symbolic links") and therefore failing to mount the OEM partition, which has recently been switched to btrfs. This is probably an issue with my local build / setup. I'll troubleshoot this and report back once I see the tests passing.

@simoncampion I think you also hit this bug https://github.com/flatcar/Flatcar/issues/1389.

ader1990 reviewed (commented):
This PR is quite heavy on the initrd size increase - ~2MB. Is it okay to merge it as is (we will again get very close to the 50% used space of the /boot partition, runway left ~2MB)?

https://github.com/flatcar/Flatcar/issues/1381

pothos commented:
Can you rebase once more? There is still a conflict with the bootengine ebuild

simoncampion commented:
@pothos I rebased again.

@ader1990 Many thanks for pointing me to that issue! I checked the logs and, indeed, before the `btrfs` error I get exactly the error you describe in the issue. I'll see whether I can make some progress fixing it.

pothos commented:
@simoncampion Thanks for working on this (much-requested!) feature. In the beginning we didn't really know how it would look and thanks to the willingness to do some extra explorations we now have something that's quite clean and also integrates well with systemd upstream tooling :)

pothos commented:
I forgot that we could have (temporarily) used the mantle PR container image ref to get things tested already. The arm64 case needs some fixes: https://github.com/flatcar/mantle/pull/509

Note that the systemd 255 update that was done in parallel also created a regression that couldn't be caught before merging because that branch wasn't using the new mantle tests yet even though it for testing it was rebased on this branch here.
